src/main/java/com/mongodb/hibernate/jdbc/MongoConnectionProvider.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2024-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.jdbc;
+
+import static org.hibernate.cfg.JdbcSettings.*;
+
+import com.mongodb.ConnectionString;
+import com.mongodb.MongoClientSettings;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import com.mongodb.hibernate.NotYetImplementedException;
+import java.sql.Connection;
+import java.util.Map;
+import org.hibernate.HibernateException;
+import org.hibernate.cfg.JdbcSettings;
+import org.hibernate.engine.jdbc.connections.spi.ConnectionProvider;
+import org.hibernate.service.UnknownUnwrapTypeException;
+import org.hibernate.service.spi.Configurable;
+import org.hibernate.service.spi.Stoppable;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * MongoDB dialect's customized JDBC {@link ConnectionProvider} SPI implementation, whose class name could be provided
+ * using configuration property {@value org.hibernate.cfg.JdbcSettings#CONNECTION_PROVIDER}.
+ *
+ * <p>The following Hibernate JDBC properties will be relied upon by Hibernate's {@link Configurable} SPI mechanism:
+ *
+ * <ul>
+ *   <li>{@linkplain org.hibernate.cfg.JdbcSettings#JAKARTA_JDBC_URL jakarta.persistence.jdbc.url}
+ *   <li>{@linkplain org.hibernate.cfg.JdbcSettings#JAKARTA_JDBC_USER jakarta.persistence.jdbc.user}
+ *   <li>{@linkplain org.hibernate.cfg.JdbcSettings#JAKARTA_JDBC_PASSWORD jakarta.persistence.jdbc.password}
+ * </ul>
+ *
+ * <p>{@value org.hibernate.cfg.JdbcSettings#JAKARTA_JDBC_URL} property is mandatory and it maps to MongoDB's <a
+ * href="https://www.mongodb.com/docs/manual/reference/connection-string/">connection string</a>, in which database name
+ * must be provided to align with JDBC URL's convention. The other two JDBC properties are optional.
+ *
+ * @see ConnectionProvider
+ * @see JdbcSettings#JAKARTA_JDBC_URL
+ * @see JdbcSettings#JAKARTA_JDBC_USER
+ * @see JdbcSettings#JAKARTA_JDBC_PASSWORD
+ * @see <a href="https://www.mongodb.com/docs/manual/reference/connection-string/">connection string</a>
+ */
+public class MongoConnectionProvider implements ConnectionProvider, Configurable, Stoppable {
```

Comment:
I've added serialVersionUID with help of the Intelllij IDE. What is the reason to use 'final', btw? I have no idea why services are all Serializable, but I am sure Hibernate codebase seldom uses serialVersionUID, but that is not a good pattern.

Reply:
> Hibernate codebase seldom uses serialVersionUID

The specification recommends to specify an explicit value:

["_It is strongly recommended that all serializable classes explicitly declare serialVersionUID values, since the default serialVersionUID computation is highly sensitive to class details that may vary depending on compiler implementations, and can thus result in unexpected serialVersionUID conflicts during deserialization, causing deserialization to fail._"](https://docs.oracle.com/en/java/javase/23/docs/specs/serialization/class.html#:~:text=It%20is%20strongly,deserialization%20to%20fail.)

> I've added serialVersionUID with help of the Intelllij IDE

Those usually huge generated values make no sense. 0 or 1 is as good, but is much more human-friendly. We started using 1 in the Java driver codebase, because ["_The serialVersionUID of a class that does not implement Serializable is 0L._"](https://docs.oracle.com/en/java/javase/23/docs/specs/serialization/class.html#:~:text=The%20serialVersionUID%20of%20a%20class%20that%20does%20not%20implement%20Serializable%20is%200L.). However, based on the new part of the spec for records, it seems that 0 is also an OK value: ["_Record classes have a default serialVersionUID value of 0L, but can declare an explicit serialVersionUID._"](https://docs.oracle.com/en/java/javase/23/docs/specs/serialization/class.html#:~:text=Record%20classes%20have%20a%20default%20serialVersionUID%20value%20of%200L%2C%20but%20can%20declare%20an%20explicit%20serialVersionUID.).

Let's use either 0 or 1. I prefer 1 to stay consistent with the Java driver codebase.

Reply:
updated by using 1